linter: `jest/no-confusing-set-timeout` reports on arbitrary `jest.*` calls when no `jest.setTimeout` exists

## What version of Oxlint are you using?

Reproduced on:

- `oxlint` 1.69.0, macOS arm64
- `oxlint` 1.60.0, macOS arm64

`npm view oxlint version` currently reports `1.69.0`.

I also checked `oxc` `main` at `f57b55db11b4bfb50179ab2ca6bce2d394032720`: the source still contains the same rule logic described in "Root cause" below. I could not run a source-built `oxlint` binary in this local environment because the repo pins Rust `1.96.0`, no Rust toolchains are installed, and both `rustup` and `mise x rust@1.96.0` fail to download from `static.rust-lang.org` with `unsuccessful tunnel`.

## What command did you run?

```sh
npx -y oxlint@1.69.0 --jest-plugin -D jest/no-confusing-set-timeout example.spec.ts
```

## What does your config look like?

No config file is needed. The repro enables the Jest plugin with `--jest-plugin` and enables only this rule with `-D jest/no-confusing-set-timeout`.

## What happened?

`jest/no-confusing-set-timeout` reports "`jest.setTimeout` should be placed before any other jest methods" on `jest.fn()` and `jest.clearAllMocks()` even though the file contains no `jest.setTimeout` call at all.

It also reports the same span multiple times. In a larger repo this produced 265,800 diagnostics across 764 files, none of which call `jest.setTimeout`.

## Minimal repro

`example.spec.ts`:

```ts
describe("example", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("works", () => {
    const fn = jest.fn();
    fn();
    expect(fn).toHaveBeenCalled();
  });
});
```

Output from `oxlint` 1.69.0:

```text
example.spec.ts:7:16: error jest(no-confusing-set-timeout): `jest.setTimeout` should be placed before any other jest methods.
example.spec.ts:3:5: error jest(no-confusing-set-timeout): `jest.setTimeout` should be placed before any other jest methods.
example.spec.ts:7:16: error jest(no-confusing-set-timeout): `jest.setTimeout` should be placed before any other jest methods.
example.spec.ts:3:5: error jest(no-confusing-set-timeout): `jest.setTimeout` should be placed before any other jest methods.
example.spec.ts:7:16: error jest(no-confusing-set-timeout): `jest.setTimeout` should be placed before any other jest methods.
```

Output from `oxlint` 1.60.0 shows the same five diagnostics, just with the older formatted reporter.

## What did you expect to happen?

Zero diagnostics. The rule should only report actual `jest.setTimeout(...)` calls that are:

- outside global scope
- repeated
- placed after other Jest methods

It should not report arbitrary `jest.*` calls as if they were `jest.setTimeout`.

## Root cause

In `crates/oxc_linter/src/rules/jest/no_confusing_set_timeout.rs` on `main` commit `f57b55db11b4bfb50179ab2ca6bce2d394032720`:

- Lines 127-145: `collect_jest_reference_id` collects every `jest.*` member expression span into `jest_reference_list`. It only checks that the reference name is `jest` and that the parent is a member expression. It does not check that the member property is `setTimeout`.
- Lines 164-168: when the rule sees any earlier non-`jest` reference that is parsed as a Jest function call, it reports every later collected `jest.*` member span with the unordered-`setTimeout` diagnostic.

That makes the rule treat later `jest.fn()`, `jest.clearAllMocks()`, etc. as if they were late `jest.setTimeout()` calls.

The duplicate diagnostics come from the same loop: one later `jest.*` span can be "later than" multiple earlier Jest method references, so the same span is reported once per earlier method.

The checked-in snapshot `crates/oxc_linter/src/snapshots/jest_no_confusing_set_timeout.snap` already contains repeated diagnostics for the same late `jest.setTimeout` span, which matches the duplicate-reporting part of this bug.

## Fix sketch

Filter the collected reference list to actual `jest.setTimeout` static member expressions, and de-duplicate unordered diagnostics by the `ReferenceId` of the offending `jest.setTimeout`.

I applied this direction locally as a candidate patch shape, but could not run the Rust test suite in this environment because of the toolchain download blocker above:

- Rename `collect_jest_reference_id` to collect only `StaticMemberExpression` nodes whose property name is `setTimeout`.
- Pass only those references to the unordered check.
- Track reported unordered `ReferenceId`s in an `FxHashSet` so one late `jest.setTimeout` produces one unordered diagnostic even if multiple earlier Jest methods exist.
- Add the minimal repro above as a passing fixture.

## Existing issue/PR search

I did not find a direct duplicate.

Related but not this bug:

- #492 is the eslint-plugin-jest umbrella/tracking issue.
- #22498 is about `--vitest-plugin` enabling Jest rules and is closed.
- Matching PRs are the original rule implementation (#938) plus later docs/perf changes (#16815, #17324), not a fix for this false positive.

## Disclosure

This was prepared with AI assistance and has been reviewed by a human.

--- a/crates/oxc_linter/src/rules/jest/no_confusing_set_timeout.rs
+++ b/crates/oxc_linter/src/rules/jest/no_confusing_set_timeout.rs
@@ -2,8 +2,8 @@ use oxc_ast::AstKind;
 use oxc_diagnostics::OxcDiagnostic;
 use oxc_macros::declare_oxc_lint;
 use oxc_semantic::{AstNode, NodeId, ReferenceId};
-use oxc_span::{GetSpan, Span};
-use rustc_hash::FxHashMap;
+use oxc_span::Span;
+use rustc_hash::{FxHashMap, FxHashSet};
 
 use crate::{
     context::LintContext,
@@ -89,6 +89,7 @@ impl Rule for NoConfusingSetTimeout {
             });
 
         let mut jest_reference_id_list: Vec<(ReferenceId, Span)> = vec![];
+        let mut reported_unordered_set_timeouts = FxHashSet::default();
         let mut seen_jest_set_timeout = false;
 
         for reference_ids in scopes.root_unresolved_references_ids() {
@@ -109,6 +110,7 @@ impl Rule for NoConfusingSetTimeout {
                 reference_id_list,
                 &jest_reference_id_list,
                 &mut seen_jest_set_timeout,
+                &mut reported_unordered_set_timeouts,
                 &id_to_jest_node_map,
             );
         }
@@ -119,6 +121,7 @@ impl Rule for NoConfusingSetTimeout {
                 reference_id_list.iter().copied(),
                 &jest_reference_id_list,
                 &mut seen_jest_set_timeout,
+                &mut reported_unordered_set_timeouts,
                 &id_to_jest_node_map,
             );
         }
@@ -139,19 +142,22 @@ fn collect_jest_reference_id(
         if !is_jest_call(ctx.semantic().reference_name(reference)) {
             continue;
         }
-        let parent_node = nodes.parent_node(reference.node_id());
-        if !parent_node.kind().is_member_expression_kind() {
+        let AstKind::StaticMemberExpression(expr) = nodes.parent_node(reference.node_id()).kind()
+        else {
             continue;
+        };
+        if expr.property.name == "setTimeout" {
+            jest_reference_list.push((reference_id, expr.span));
         }
-        jest_reference_list.push((reference_id, parent_node.kind().span()));
     }
 }
 
 fn handle_jest_set_time_out<'a>(
     ctx: &LintContext<'a>,
     reference_id_list: impl Iterator<Item = ReferenceId>,
-    jest_reference_id_list: &Vec<(ReferenceId, Span)>,
+    jest_reference_id_list: &[(ReferenceId, Span)],
     seen_jest_set_timeout: &mut bool,
+    reported_unordered_set_timeouts: &mut FxHashSet<ReferenceId>,
     id_to_jest_node_map: &FxHashMap<NodeId, &PossibleJestNode<'a, '_>>,
 ) {
     let nodes = ctx.nodes();
@@ -165,7 +171,9 @@ fn handle_jest_set_time_out<'a>(
         if !is_jest_call(ctx.semantic().reference_name(reference)) {
             if is_jest_fn_call(parent_node, id_to_jest_node_map, ctx) {
                 for (jest_reference_id, span) in jest_reference_id_list {
-                    if jest_reference_id > &reference_id {
+                    if jest_reference_id > &reference_id
+                        && reported_unordered_set_timeouts.insert(*jest_reference_id)
+                    {
                         ctx.diagnostic(no_unorder_set_timeout_diagnostic(*span));
                     }
                 }
@@ -345,6 +353,22 @@ fn test() {
                     setTimeout(() => {
                         Promise.resolv();
                     }, 5000);
+                });
+            ",
+            None,
+        ),
+        (
+            "
+                describe('example', () => {
+                    beforeEach(() => {
+                        jest.clearAllMocks();
+                    });
+
+                    it('works', () => {
+                        const fn = jest.fn();
+                        fn();
+                        expect(fn).toHaveBeenCalled();
+                    });
                 });
             ",
             None,

